packages/twenty-server/src/engine/metadata-modules/agent/agent-tool.service.ts

```diff
@@ -0,0 +1,827 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+
+import { ToolSet } from 'ai';
+import {
+  In,
+  IsNull,
+  LessThan,
+  LessThanOrEqual,
+  Like,
+  MoreThan,
+  MoreThanOrEqual,
+  Not,
+  Repository,
+} from 'typeorm';
+import { z } from 'zod';
+
+import { DatabaseEventAction } from 'src/engine/api/graphql/graphql-query-runner/enums/database-event-action';
+import { AgentService } from 'src/engine/metadata-modules/agent/agent.service';
+import { ObjectMetadataService } from 'src/engine/metadata-modules/object-metadata/object-metadata.service';
+import { RoleEntity } from 'src/engine/metadata-modules/role/role.entity';
+import { TwentyORMGlobalManager } from 'src/engine/twenty-orm/twenty-orm-global.manager';
+import { WorkspaceEventEmitter } from 'src/engine/workspace-event-emitter/workspace-event-emitter';
+
+import {
+  generateAgentToolUpdateZodSchema,
+  generateAgentToolZodSchema,
+  generateBulkDeleteToolSchema,
+  generateFindToolSchema,
+} from './utils/agent-tool-schema.utils';
+import { isWorkflowRelatedObject } from './utils/is-workflow-related-object.util';
+
+@Injectable()
+export class AgentToolService {
+  constructor(
+    private readonly agentService: AgentService,
+    @InjectRepository(RoleEntity, 'core')
+    private readonly roleRepository: Repository<RoleEntity>,
+    private readonly objectMetadataService: ObjectMetadataService,
+    private readonly twentyORMGlobalManager: TwentyORMGlobalManager,
+    private readonly workspaceEventEmitter: WorkspaceEventEmitter,
+  ) {}
+
+  async generateToolsForAgent(
+    agentId: string,
+    workspaceId: string,
+  ): Promise<ToolSet> {
+    try {
+      const agent = await this.agentService.findOneAgent(agentId, workspaceId);
+
+      if (!agent.roleId) {
+        return {};
+      }
+
+      const roleWithPermissions = await this.roleRepository.findOne({
+        where: {
+          id: agent.roleId,
+          workspaceId,
+        },
+        relations: [
+          'objectPermissions',
+          'objectPermissions.objectMetadata',
+          'objectPermissions.objectMetadata.fields',
+        ],
+      });
+
+      if (!roleWithPermissions) {
+        return {};
+      }
+
+      const tools: ToolSet = {};
+
+      const allObjectMetadata =
+        await this.objectMetadataService.findManyWithinWorkspace(workspaceId, {
+          where: {
+            isActive: true,
+            isSystem: false,
+          },
+          relations: ['fields'],
+        });
+
+      const filteredObjectMetadata = allObjectMetadata.filter(
+        (objectMetadata) => !isWorkflowRelatedObject(objectMetadata),
+      );
+
+      const objectPermissionsMap = new Map(
```

Comment:
I don't have enough knowledge of this part of the codebase but this feels like something that could be re-used from elsewhere? cc @ijreilly if you have an input (we are rebuilding object permission map from scratch!).
In the future it will get more complex with things like field level-permissions so you @abdulrahmancodes you'll want to use the same service to make sure we don't have that logic is in two different places. There is probably some caching mechanism in @ijreilly's work that you would benefit from